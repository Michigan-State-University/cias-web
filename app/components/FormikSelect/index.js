/**
 *
 * FormikSelect
 *
 */

import React, { memo } from 'react';
import PropTypes from 'prop-types';
import { useField, useFormikContext } from 'formik';
import isNil from 'lodash/isNil';

import Select from 'components/Select';
import FormikControlLayout from 'components/FormikControlLayout';

function FormikSelect({
  formikKey,
  label,
  options,
  columnStyleProps,
  inputProps,
  disabled,
  submitOnChange,
}) {
  const { submitForm, validateForm } = useFormikContext();
  const [field, meta, helpers] = useField(formikKey);
  const { error, touched } = meta;
  const { setValue } = helpers;
  const hasError = touched && !isNil(error);

  const onChange = async (e) => {
    if (submitOnChange) {
      await validateForm();
      setValue(e);
      await submitForm();
    } else {
      setValue(e);
    }
  };

  return (
    <FormikControlLayout
      formikKey={formikKey}
      label={label}
      touched={touched}
      error={error}
      {...columnStyleProps}
    >
      <Select
        isOptionDisabled={disabled}
        data-testid="select"
        disabled={disabled}
        selectProps={{
          options,
          isDisabled: disabled,
          inputId: formikKey,
          hasError,
          ...field,
          onChange,
          ...inputProps,
        }}
      />
    </FormikControlLayout>
  );
}

FormikSelect.propTypes = {
  formikKey: PropTypes.string.isRequired,
  label: PropTypes.string,
  inputProps: PropTypes.object,
  options: PropTypes.array,
  columnStyleProps: PropTypes.object,
  disabled: PropTypes.bool,
  submitOnChange: PropTypes.bool,
};

export default memo(FormikSelect);
