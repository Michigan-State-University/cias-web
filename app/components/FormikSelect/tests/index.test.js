/**
 *
 * Tests for FormikSelect
 *
 */

import React, { Fragment } from 'react';
import { render } from '@testing-library/react';
import { Formik } from 'formik';

import FormikSelect from '../index';

describe('<FormikSelect />', () => {
  let props;
  beforeAll(() => {
    props = {
      formikKey: 'test',
      placeholder: 'test',
      label: 'Label',
      options: [{ label: 'option', value: 'option' }],
    };
  });
  it('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');
    render(
      <Formik
        initialValues={{ test: { label: 'option', value: 'option' } }}
        onSubmit={jest.fn()}
      >
        {() => (
          <Fragment>
            <FormikSelect {...props} />
          </Fragment>
        )}
      </Formik>,
    );
    expect(spy).not.toHaveBeenCalled();
  });

  it('Should render and match the snapshot', () => {
    const { container } = render(
      <Formik
        initialValues={{ test: { label: 'option', value: 'option' } }}
        onSubmit={jest.fn()}
      >
        {() => (
          <Fragment>
            <FormikSelect {...props} />
          </Fragment>
        )}
      </Formik>,
    );
    expect(container).toMatchSnapshot();
  });
});
