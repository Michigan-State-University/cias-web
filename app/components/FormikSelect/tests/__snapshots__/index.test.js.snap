// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormikSelect /> Should render and match the snapshot 1`] = `
.c2 {
  width: auto;
  height: auto;
  display: block;
  border-radius: 5px;
  cursor: pointer;
}

.c3 {
  width: auto;
  height: auto;
  -webkit-transition: -webkit-transform 0.2s;
  -webkit-transition: transform 0.2s;
  transition: transform 0.2s;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
}

.c1 {
  font-family: DM Sans;
  font-size: 13px;
  line-height: 130%;
  margin-bottom: 5px;
  color: #2F3850;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
}

@media only screen and (min-width:768px) {

}

@media only screen and (min-width:960px) {

}

@media only screen and (min-width:1170px) {

}

@media only screen and (min-width:1360px) {

}

<div>
  <div
    class="c0"
  >
    <label
      for="test"
    >
      <p
        class="c1"
        width="fit-content"
      >
        Label
      </p>
    </label>
    <div
      class="c2"
      cursor="pointer"
      data-testid="select"
    >
      <div
        class=" css-2b097c-container"
      >
        <span
          aria-atomic="false"
          aria-live="polite"
          aria-relevant="additions text"
          class="css-1f43avz-a11yText-A11yText"
        />
        <div
          class=" css-4iyz22-control"
        >
          <div
            class=" css-g1d714-ValueContainer"
          >
            <div
              class=" css-1uccc91-singleValue"
            >
              option
            </div>
            <div
              class="css-b8ldur-Input"
            >
              <div
                class=""
                style="display: inline-block;"
              >
                <input
                  aria-autocomplete="list"
                  autocapitalize="none"
                  autocomplete="off"
                  autocorrect="off"
                  id="test"
                  spellcheck="false"
                  style="box-sizing: content-box; width: 2px; border: 0px; opacity: 1; outline: 0; padding: 0px;"
                  tabindex="0"
                  type="text"
                  value=""
                />
                <div
                  style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                />
              </div>
            </div>
          </div>
          <div
            class=" css-1hb7zxy-IndicatorsContainer"
          >
            <div
              aria-hidden="true"
              class=" css-tlfecz-indicatorContainer"
            >
              <img
                alt="arrow"
                class="c3"
                src="IMAGE_MOCK"
                transform=""
              />
            </div>
          </div>
        </div>
        <input
          name="test"
          type="hidden"
          value="option"
        />
      </div>
    </div>
  </div>
</div>
`;
